File: part-01/src/main/java/eu/luminis/workshop/smallsteps/logic/UserService.java
package eu.luminis.workshop.smallsteps.logic;

import eu.luminis.workshop.smallsteps.api.NewUserRequest;
import eu.luminis.workshop.smallsteps.persistence.UserPersistence;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

@Service
public class UserService {
    @Autowired
    UserPersistence userPersistence;

    public void registerNewUser(NewUserRequest request) {
        if (!StringUtils.hasText(request.getEmail())) {
            throw new IllegalArgumentException("Please enter an email address");
        }
        if (!StringUtils.hasText(request.getPassword()) || request.getPassword().length() < 8) {
            throw new IllegalArgumentException("Password must be at least 8 characters long");
        }

        userPersistence.insertUser(request);
    }
}

File: part-01/src/main/kotlin/eu/luminis/workshop/smallsteps/logic/UserService.kt
package eu.luminis.workshop.smallsteps.logic

import java.util.*

// Understands how users can be registered
class UserService(private val userPersistence: UserPersistence) {
    suspend fun registerNewUser(command: NewUserCommand) {
        userPersistence.insertUser(command.email, command.password)
    }
}

data class NewUserCommand(
    val email: Email,
    val password: Password,
)

interface UserPersistence {
    suspend fun insertUser(email: Email, password: Password): UUID
}